codegen/module.go

```diff
@@ -104,6 +90,79 @@ func (system *ModuleSystem) RegisterClass(class ModuleClass) error {
 	return nil
 }
 
+// dedup returns a slice containing unique sorted elements of the given slice
+func dedup(array []string) []string {
+	dict := map[string]bool{}
+	for _, elt := range array {
+		dict[elt] = true
+	}
+
+	i := 0
+	unique := make([]string, len(dict))
+	for key := range dict {
+		unique[i] = key
+		i++
+	}
+	sort.Strings(unique)
+	return unique
+}
+
+// validateDir checks if the module class can map to the given dir
+func (system *ModuleSystem) validateClassDir(class *ModuleClass, dir string) error {
+	dir = filepath.Clean(dir)
+
+	if strings.HasPrefix(dir, "..") {
+		return errors.Errorf(
+			"Module class %q must map to internal directories but found %q",
+			class.Name,
+			dir,
+		)
+	}
+
+	// Validate the module class directories are unique
```

Comment:
It might not be strictly necessary to enforce this requirement. It would be possible for two class types to live in the same directory, as the instances are distinguishable by their config name.